iroh/src/magicsock/relay_actor.rs

```diff
@@ -964,4 +951,163 @@ mod tests {
+        let echo_task = AbortOnDropHandle::new(echo_task);
+        let supervisor_task = tokio::spawn(async move {
+            // move the inbox_tx here so it is not dropped, as this stops the actor.
+            let _inbox_tx = inbox_tx;
+            tokio::select! {
+                biased;
+                _ = actor_task => (),
+                _ = echo_task => (),
+            };
+        });
+        let supervisor_task = AbortOnDropHandle::new(supervisor_task);
+        (secret_key.public(), supervisor_task)
+    }
+
+    #[tokio::test]
+    async fn test_active_relay_reconnect() -> TestResult {
+        let _guard = iroh_test::logging::setup();
+        let (_relay_map, relay_url, _server) = test_utils::run_relay_server().await?;
+        let (peer_node, _echo_node_task) = start_echo_node(relay_url.clone());
+
+        let secret_key = SecretKey::from_bytes(&[1u8; 32]);
+        let datagram_recv_queue = Arc::new(RelayDatagramsQueue::new());
+        let (send_datagram_tx, send_datagram_rx) = mpsc::channel(16);
+        let (inbox_tx, inbox_rx) = mpsc::channel(16);
+        let task = start_active_relay_actor(
+            secret_key,
+            relay_url,
+            inbox_rx,
+            send_datagram_rx,
+            datagram_recv_queue.clone(),
+        );
+
+        // Send a datagram to our echo node.
+        info!("first echo");
+        let packet = PacketizeIter::<_, MAX_PAYLOAD_SIZE>::new(peer_node, [b"hello"])
+            .next()
+            .context("no packet")?;
+        send_datagram_tx.send(packet).await?;
+
+        // Check we get it back
+        let RelayRecvDatagram {
+            url: _,
+            src: _,
+            buf,
+        } = future::poll_fn(|cx| datagram_recv_queue.poll_recv(cx)).await?;
+        assert_eq!(buf.as_ref(), b"hello");
+
+        // Now ask to check the connection, triggering a ping but no reconnect.
+        let (tx, rx) = oneshot::channel();
+        inbox_tx.send(ActiveRelayMessage::GetLocalAddr(tx)).await?;
+        let local_addr = rx.await?.context("no local addr")?;
+        info!(?local_addr, "check connection with addr");
+        inbox_tx
+            .send(ActiveRelayMessage::CheckConnection(vec![local_addr.ip()]))
+            .await?;
+
+        // Sync the ActiveRelayActor. Ping blocks it and we want to be sure it has handled
+        // another inbox message before continuing.
+        let (tx, rx) = oneshot::channel();
+        inbox_tx.send(ActiveRelayMessage::GetLocalAddr(tx)).await?;
+        rx.await?;
+
+        // Echo should still work.
+        info!("second echo");
+        let packet = PacketizeIter::<_, MAX_PAYLOAD_SIZE>::new(peer_node, [b"hello"])
+            .next()
+            .context("no packet")?;
+        send_datagram_tx.send(packet).await?;
+        let recv = future::poll_fn(|cx| datagram_recv_queue.poll_recv(cx)).await?;
+        assert_eq!(recv.buf.as_ref(), b"hello");
+
+        // Now ask to check the connection, this will reconnect.
```

Comment:
why does this reconnect? when was the connection dropped?

Reply:
Because the `CheckConnection` message does not contain any local IPs the ActiveRelayActor will consider the local addr of the connection to be no longer available on the host so automatically reconnect without first pinging.

The ActiveRelayActor itself now handles the reconnect (in the new `reconnect` method) and simply drops the previous client.

What actually happens is that the drop might not happen soon enough, depending on async runtime scheduling, and the relay server sees two connections and closes the older one.  But sometimes the drop happens first and the relay connection sees the connection being closed (abruptly).

Reply:
maybe add a short comment so
we don’t forget